dwave/optimization/generators.py

```diff
@@ -145,6 +146,69 @@ def quadratic_model(qm):
 
     return model
 
+def bin_packing(weights: numpy.typing.ArrayLike,
+                bin_capacity: float
+                ) -> Model:
+    r"""Generate a model encoding a bin packing problem.
+
+    The bin packing problem,
+    `BPP <https://en.wikipedia.org/wiki/Bin_packing_problem>`_,
+    seeks to find the smallest number of bins that will fit
+    a set of weighted items given that each bin has a weight capacity.
+
+    Args:
+        weights:
+            A 1D |array-like|_ (row vector or Python list) of weights per item.
+            Weights can be any non-negative number.
+
+        bin_capacity:
+            Maximum capacity of each bin. Must be a positive number.
+
+    Returns:
+        A model encoding the BPP problem.
+    """
+
+    weights = _require("weights", weights, dtype=float, nonnegative=True, ndim=1)
+
+    if weights.shape[0] == 0:
+        raise ValueError("the number of items must be positive")
+
+    if bin_capacity <= 0:
+        raise ValueError("`bin_capacity` must be a positive number")
+
+    if any(weight > bin_capacity for weight in weights):
+        raise ValueError("each weight in `weights` must be smaller than bin_capacity")
+
+    num_items = weights.shape[0]
+
+    # Construct model
+    model = Model()
+
+    # Add constants
+    weights = model.constant(weights)
+    bin_capacity = model.constant(bin_capacity)
+
+    # Create disjoint bit sets to represent bins
+    max_bins = num_items
+    main, bins = model.disjoint_bit_sets(num_items, max_bins)
```

Comment:
This could also potentially be formulated with `DisjointSets`. The main benefit there is we could use the `Size` symbol (added in https://github.com/dwavesystems/dwave-optimization/pull/50) to determine set emptiness. Not sure about the performance implications though.

Reply:
Do you mean `DisjointLists`?

Reply:
Oh, I forgot that doesn't exist yet. It's to `DisjointLists` what `Set` is to `List`. But I guess we haven't implemented it 😳

Reply:
Okay, should I wait until it exists, or use `DisjointBitSets` (and maybe update when `DisjointSets` exists), or use `DisjointLists` (probably not that last one)?

Reply:
I think it makes sense to stick with `DisjointBitSets` for now, and update later if we like.